### ISSUE TITLE
Why getClientInstance is used for inbound and getAsyncClientInstance is used for outbound to get synchronous and asynchronous clients. 

### ISSUE THREAD

**Issue by qq2429979710** [spring-projects/spring-integration-samples#331]
What if I use an asynchronous client for both outbound and inbound, I see that the code is private and there is no way to rewrite it 

`
this.client = this.clientFactory.getAsyncClientInstance(this.getUrl(), this.getClientId());
`

**Comment by garyrussell** [comment]
It is not clear what you mean; please elaborate. 

We use the async client in all places. 

It will block (outbound) to wait for the result, unless you set `async` to true.

**Comment by qq2429979710** [comment]
Here is the code snippet：

```
MqttPahoClientFactory     interface

	IMqttClient getClientInstance(String url, String clientId) throws MqttException;

	IMqttAsyncClient getAsyncClientInstance(String url, String clientId) throws MqttException;

Inbound：
  connectAndSubscribe   method
  this.client = this.clientFactory.getClientInstance(getUrl(), getClientId());
Outbound：
  checkConnection  method
  this.client = this.clientFactory.getAsyncClientInstance(this.getUrl(), this.getClientId());
```

I see different clients used for inbound and outbound creation clients


What I want to do is to make sure that the inbound and outbound are using the same client by rewriting MqttPahoClientFactory to get the client, but I can cache it against the HashMap

like this :

```
   private Map<String,IMqttAsyncClient> iMqttAsyncClientMap = new ConcurrentHashMap<>();
    @Override
    public IMqttAsyncClient getAsyncClientInstance(String uri, String clientId) throws MqttException {
        String key = uri + clientId;
        IMqttAsyncClient iMqttAsyncClient = iMqttAsyncClientMap.get(key);
        if(iMqttAsyncClient==null){
            iMqttAsyncClient= super.getAsyncClientInstance(uri, clientId);
            iMqttAsyncClientMap.put(key,iMqttAsyncClient);
        }
        return iMqttAsyncClient;
    }
```
However, I found that the inbound and outbound terminals were not the same client, so my idea could not be implemented

**Comment by garyrussell** [comment]
There is a recent (merged) pull request for sharing clients across components https://github.com/spring-projects/spring-integration/pull/3857 for the upcoming 6.0 release.

In previous versions, a sync Vs async made no different on the inbound side.

**Comment by qq2429979710** [comment]
Ok, my requirement is to use the same clientID client and push and subscribe.

**Comment by artembilan** [comment]
Right, we didn't support shared client between channel adapter until the fix Gary has shared with you.

It is probably possible to implement a custom `MqttPahoClientFactory` to return the same client for every `getAsyncClientInstance()` call.
The `getClientInstance()` can return a custom (and same) `IMqttClient` which can delegate to respective `IMqttAsyncClient`. That's essentially what is going on in the `org.eclipse.paho.client.mqttv3.MqttClient`.

I'm treating this issue as `Invalid` since this project is not about asking questions.
Ask them on StackOverflow, please.
